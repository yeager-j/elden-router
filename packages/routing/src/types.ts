import { MultiGraph } from "graphology";

import { Enemy } from "@workspace/data/enemies";
import { Flag } from "@workspace/data/flags";
import { Glitch } from "@workspace/data/glitches";
import { Item, ProgressionItem } from "@workspace/data/items";
import { Location } from "@workspace/data/locations";
import { QuestlineStage } from "@workspace/data/quests";

export enum BossPreference {
  NONE = "NONE",
  MINIMAL = "MINIMAL",
  ANY = "ANY",
}

interface BossPreferenceMetadata {
  displayName: string;
  description: string;
}

export const BossPreferenceData: Record<
  BossPreference,
  BossPreferenceMetadata
> = {
  [BossPreference.NONE]: {
    displayName: "None",
    description: "Do not allow bosses in the route",
  },
  [BossPreference.MINIMAL]: {
    displayName: "Minimal",
    description: "Find a route with the minimal number of bosses",
  },
  [BossPreference.ANY]: {
    displayName: "Any",
    description: "Find the shortest route regardless of bosses",
  },
};

export type MultiLocationItemNode = string & {
  __brand: "MultiLocationItemNode";
};

export type EldenGraph = MultiGraph<object, EdgeMetadata>;

export type GraphNode = Location | Item | MultiLocationItemNode;

export interface GetPathResult {
  pathSteps: PathStep[] | null;

  isError: boolean;
  reasons: string[];
}

export interface PathStep {
  from: Location;
  to: Location | Item | MultiLocationItemNode;
  metadata: EdgeMetadata;
}

export interface PathSettings {
  bossPreference: BossPreference;
  allowGlitches: boolean;
  acquiredItems: Set<ProgressionItem>;
  completedQuestlineStages: Set<QuestlineStage>;
  flagsEnabled: Set<Flag>;
}

export interface EdgeMetadata {
  description?: string;
  requirements: {
    requiredBosses?: Enemy[];
    requiredGlitch?: {
      glitch: Glitch;
      description: string;
    };
    requiredItems?: ProgressionItem[];
    requiredQuests?: QuestlineStage[];
    requiredEnabledFlags?: Flag[];
    requiredDisabledFlags?: Flag[];
  };
}

export interface EdgeData {
  from: Location;
  to: Location | Item | MultiLocationItemNode;
  directed?: boolean;
  metadata: EdgeMetadata;
}
