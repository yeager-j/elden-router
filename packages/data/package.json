{
  "name": "@workspace/data",
  "type": "module",
  "private": true,
  "scripts": {
    "dev": "tsc --watch",
    "build": "tsc"
  },
  "exports": {
    "./enemies": "./src/enemies.ts",
    "./flags": "./src/flags.ts",
    "./glitches": "./src/glitches.ts",
    "./items": "./src/items.ts",
    "./locations": "./src/locations.ts",
    "./quests": "./src/quests.ts"
  },
  "imports": {
    "#*": [
      "./src/*.ts"
    ]
  },
  "prettier": "@workspace/prettier-config",
  "devDependencies": {
    "@workspace/eslint-config": "workspace:*",
    "@workspace/prettier-config": "workspace:*",
    "@workspace/typescript-config": "workspace:*",
    "@vitest/coverage-v8": "^3.0.5",
    "typescript": "^5.6.3",
    "vite": "^6.0.11",
    "vite-tsconfig-paths": "^5.1.4"
  },
  "dependencies": {
    "graphology": "^0.26.0",
    "graphology-shortest-path": "^2.1.0"
  }
}