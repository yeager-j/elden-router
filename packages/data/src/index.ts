export * from "./enemies";
export * from "./flags";
export * from "./glitches";
export * from "./items";
export * from "./locations";
export * from "./quests";
