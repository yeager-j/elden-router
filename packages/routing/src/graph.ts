import { MultiGraph } from "graphology";
import { dijkstra } from "graphology-shortest-path";

import { Item } from "@workspace/data/items";
import { Location } from "@workspace/data/locations";

import { allEdges } from "#edges";
import {
  BossPreference,
  EldenGraph,
  GetPathResult,
  PathSettings,
  PathStep,
} from "#types";
import {
  getPartsFromMultiLocationItem,
  isMultiLocationItemNode,
} from "#utils/edge-utils";
import { checkEdgeMeetsSettings, getBestEdge } from "#utils/graph-utils";

type Destination = Item | Location;

/**
 * Retrieves a list of possible destination items by analyzing a graph and identifying specific nodes.
 *
 * @return {Item[]} An array of unique items that represent possible destinations extracted from the graph.
 */
export function getPossibleDestinations(): Item[] {
  const graph = buildGraph();

  const items = new Set<Item>();

  graph.nodes().forEach((node) => {
    if (isMultiLocationItemNode(node)) {
      const [item] = getPartsFromMultiLocationItem(node);

      if (Object.values(Item).includes(item)) {
        items.add(item);
      }
    }

    if (Object.values(Item).includes(node as Item)) {
      items.add(node as Item);
    }
  });

  return Array.from(items);
}

/**
 * Determines the optimal path to a specified destination based on the provided settings.
 *
 * @param {Destination} destination - The destination to which the path is being calculated.
 * @param {PathSettings} settings - The configuration settings that influence path calculation, such as preferences and constraints.
 * @return {GetPathResult} Returns the result of the pathfinding operation, which includes the calculated path or an error reason if no suitable path exists.
 */
export function getPathToDestination(
  destination: Destination,
  settings: PathSettings,
): GetPathResult {
  const graph = buildGraph();
  // Copy the graph so we can filter edges
  const filteredGraph = graph.copy();

  // Keep track of why each edge was removed
  const removalReasons = new Map<string, string>();

  // Filter edges based on settings
  filteredGraph.forEachEdge((edge, attributes) => {
    const { valid, reason } = checkEdgeMeetsSettings(settings, attributes);

    if (!valid) {
      if (reason) {
        removalReasons.set(edge, reason);
      }

      filteredGraph.dropEdge(edge);
    }
  });

  const bestPath = findShortestPathToLocation(
    filteredGraph,
    destination,
    settings.bossPreference,
  );

  if (!bestPath) {
    // Path doesn't exist
    return constructErrorReason(
      graph,
      filteredGraph,
      removalReasons,
      destination,
    );
  }

  // If we have a path in the filtered graph, build the path steps
  return constructPath(bestPath, filteredGraph);
}

/**
 * Constructs and returns a graph based on a set of edges, adding nodes and edges appropriately.
 *
 * This method initializes a new graph and processes a collection of edges to determine the
 * nodes and connections between them. Nodes are added to the graph, and edges are created
 * as directed or undirected based on the edge metadata.
 *
 * @return {EldenGraph} A graph containing all nodes and edges as specified by the input data.
 */
function buildGraph(): EldenGraph {
  const graph: EldenGraph = new MultiGraph();

  const nodes = new Set<string>();
  allEdges.forEach((edge) => {
    nodes.add(edge.from);
    nodes.add(edge.to);
  });

  // Add all nodes to the graph
  nodes.forEach((node) => {
    graph.addNode(node);
  });

  // Add edges
  allEdges.forEach((edge) => {
    const { from, to, directed, metadata } = edge;

    if (directed) {
      graph.addDirectedEdge(from, to, metadata);
    } else {
      graph.addUndirectedEdge(from, to, metadata);
    }
  });

  return graph;
}

/**
 * Finds the shortest path to a specified destination node within the given graph.
 *
 * @param {EldenGraph} graph - The graph representing the interconnected locations and paths.
 * @param {Destination} destination - The target destination to which the shortest path is to be found.
 * @param {BossPreference} bossPreference - The preferred boss-related criteria to influence the pathfinding cost calculation.
 * @return {string[] | null} - The shortest path as an array of node names, or null if no path exists.
 */
function findShortestPathToLocation(
  graph: EldenGraph,
  destination: Destination,
  bossPreference: BossPreference,
): string[] | null {
  const destinationNodes = graph.nodes().filter((node) => {
    if (isMultiLocationItemNode(node)) {
      const [item] = getPartsFromMultiLocationItem(node);
      return item === destination;
    }

    return node === destination;
  });

  let bestPath: string[] | null = null;
  let bestCost = Infinity;

  for (const destinationNode of destinationNodes) {
    const path = dijkstra.bidirectional(
      graph,
      Location.LIMGRAVE,
      destinationNode,
      (_, attributes) => {
        if (bossPreference === BossPreference.ANY) {
          return 0;
        }

        return (attributes.requirements.requiredBosses?.length ?? 0) * 100;
      },
    );

    if (path) {
      const cost = path.reduce((sum, node, i) => {
        if (i === 0) return sum;
        return sum + 1;
      }, 0);

      if (cost < bestCost) {
        bestCost = cost;
        bestPath = path;
      }
    }
  }

  return bestPath;
}

/**
 * Constructs a path object containing steps and metadata based on the provided path and graph.
 *
 * @param {string[]} path - An array of strings representing the ordered nodes in the path.
 * @param {EldenGraph} filteredGraph - The graph object containing nodes and edges with associated attributes.
 * @return {GetPathResult} An object containing the constructed path steps, an error flag, and a list of reasons for errors if any.
 */
function constructPath(
  path: string[],
  filteredGraph: EldenGraph,
): GetPathResult {
  const pathSteps: PathStep[] = [];

  for (let i = 0; i < path.length - 1; i++) {
    const fromNode = path[i]! as Location;
    const toNode = path[i + 1]! as Location | Item;
    const edgeKeys = filteredGraph.edges(fromNode, toNode);

    const metadata = filteredGraph.getEdgeAttributes(
      getBestEdge(filteredGraph, edgeKeys),
    );

    pathSteps.push({
      from: fromNode,
      to: toNode,
      metadata,
    });
  }

  return {
    pathSteps,
    isError: false,
    reasons: [],
  };
}

/**
 * Constructs an error reason based on the provided graph, filtered graph, removal reasons, and destination.
 * It identifies why a path cannot be constructed, highlighting restrictions or missing connections.
 *
 * @param {EldenGraph} graph The complete graph containing all nodes and edges.
 * @param {EldenGraph} filteredGraph The graph with certain edges or nodes removed due to constraints.
 * @param {Map<string, string>} removalReasons A map of edges to their associated removal reasons (e.g., restrictions or requirements).
 * @param {Location | Item} destination The target destination for which the path is to be calculated.
 * @return {GetPathResult} An object containing the error state, an optional path (null in error state),
 *                          and a list of reasons explaining why no path exists or is valid.
 */
function constructErrorReason(
  graph: EldenGraph,
  filteredGraph: EldenGraph,
  removalReasons: Map<string, string>,
  destination: Location | Item,
): GetPathResult {
  const unfilteredPath = findShortestPathToLocation(
    graph,
    destination,
    BossPreference.ANY,
  );

  if (!unfilteredPath) {
    // There's no path in the unfiltered graph either.
    return {
      pathSteps: null,
      isError: true,
      reasons: ["No route exists, even ignoring requirements."],
    };
  } else {
    // The unfiltered path exists, so let's see which edges were removed.
    const reasonsSet = new Set<string>();

    for (let i = 0; i < unfilteredPath.length - 1; i++) {
      const fromNode = unfilteredPath[i] as string;
      const toNode = unfilteredPath[i + 1] as string;
      const edgeKeys = graph.edges(fromNode, toNode);

      if (!edgeKeys.length) continue;

      edgeKeys.forEach((key) => {
        // If the edge was removed, add its removal reason
        if (!filteredGraph.hasEdge(key) && removalReasons.has(key)) {
          reasonsSet.add(removalReasons.get(key)!);
        }
      });
    }

    if (reasonsSet.size === 0) {
      // If no specific reasons found, fallback message
      reasonsSet.add(
        "No path found, but a route exists if you remove some restrictions.",
      );
    }

    return {
      pathSteps: null,
      isError: true,
      reasons: Array.from(reasonsSet),
    };
  }
}
