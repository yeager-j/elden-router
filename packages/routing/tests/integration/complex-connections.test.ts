import { describe, expect, test } from "vitest";

import { Flag, Item, Location, ProgressionItem } from "@workspace/data";

import { getPathToDestination } from "@/graph";
import { PathSettings } from "@/types";

describe("complex connections", () => {
  const config: PathSettings = {
    allowBosses: true,
    allowGlitches: false,
    acquiredItems: new Set([
      ProgressionItem.DECTUS_MEDALLION,
      ProgressionItem.ROLD_MEDALLION,
      ProgressionItem.HALIGTREE_MEDALLION,
    ]),
    completedQuestlineStages: new Set(),
    flagsEnabled: new Set(),
  };

  test("Limgrave -> Somberstone Miner's Bell Bearing (5)", () => {
    const route = getPathToDestination(
      Item.SOMBERSTONE_MINERS_BELL_BEARING_5,
      config,
    );

    expect(route).toMatchSnapshot();
  });

  test("Limgrave -> Triple Rings of Light", () => {
    const route = getPathToDestination(Item.TRIPLE_RINGS_OF_LIGHT, {
      ...config,
      allowGlitches: true,
    });

    expect(route).toMatchSnapshot();
  });

  test("Limgrave -> Fractured Marika", () => {
    const route = getPathToDestination(Location.LEYNDELL_FRACTURED_MARIKA, {
      ...config,
      flagsEnabled: new Set([Flag.LEYNDELL_CAPITAL_ASHEN]),
    });

    expect(route).toMatchSnapshot();
  });

  test("Limgrave -> Ashen Leyndell -> Rivers of Blood", () => {
    const route = getPathToDestination(Item.RIVERS_OF_BLOOD, {
      ...config,
      flagsEnabled: new Set([Flag.LEYNDELL_CAPITAL_ASHEN]),
    });

    expect(route).toMatchSnapshot();
  });
});
