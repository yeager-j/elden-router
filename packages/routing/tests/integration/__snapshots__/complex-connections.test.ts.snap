// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`complex connections > Limgrave -> Ashen Leyndell -> Rivers of Blood 1`] = `
{
  "isError": false,
  "pathSteps": [
    {
      "from": "LIMGRAVE",
      "metadata": {
        "description": "Take the shortcut near Stormveil Castle",
        "requirements": {},
      },
      "to": "LIURNIA",
    },
    {
      "from": "LIURNIA",
      "metadata": {
        "description": "Access Altus Plateau via Grand Lift of Dectus",
        "requirements": {
          "requiredItems": [
            "DECTUS_MEDALLION",
          ],
        },
      },
      "to": "ALTUS_PLATEAU",
    },
    {
      "from": "ALTUS_PLATEAU",
      "metadata": {
        "requirements": {},
      },
      "to": "ALTUS_PLATEAU_LEYNDELL_OUTSKIRTS",
    },
    {
      "from": "ALTUS_PLATEAU_LEYNDELL_OUTSKIRTS",
      "metadata": {
        "description": "Leyndell turns to ash after defeating Maliketh",
        "requirements": {
          "requiredEnabledFlags": [
            "LEYNDELL_CAPITAL_ASHEN",
          ],
        },
      },
      "to": "LEYNDELL_ASHEN_CAPITAL",
    },
    {
      "from": "LEYNDELL_ASHEN_CAPITAL",
      "metadata": {
        "requirements": {
          "requiredEnabledFlags": [
            "LEYNDELL_CAPITAL_ASHEN",
          ],
        },
      },
      "to": "ALTUS_PLATEAU_FORBIDDEN_LANDS",
    },
    {
      "from": "ALTUS_PLATEAU_FORBIDDEN_LANDS",
      "metadata": {
        "description": "Access via Grand Lift of Rold",
        "requirements": {
          "requiredItems": [
            "ROLD_MEDALLION",
          ],
        },
      },
      "to": "MOUNTAINTOPS_OF_THE_GIANTS",
    },
    {
      "from": "MOUNTAINTOPS_OF_THE_GIANTS",
      "metadata": {
        "description": "Defeat Bloody Finger Okina when he invades in the Church of Repose",
        "requirements": {},
      },
      "to": "RIVERS_OF_BLOOD",
    },
  ],
  "reasons": [],
}
`;

exports[`complex connections > Limgrave -> Fractured Marika 1`] = `
{
  "isError": false,
  "pathSteps": [
    {
      "from": "LIMGRAVE",
      "metadata": {
        "description": "Take the shortcut near Stormveil Castle",
        "requirements": {},
      },
      "to": "LIURNIA",
    },
    {
      "from": "LIURNIA",
      "metadata": {
        "description": "Access Altus Plateau via Grand Lift of Dectus",
        "requirements": {
          "requiredItems": [
            "DECTUS_MEDALLION",
          ],
        },
      },
      "to": "ALTUS_PLATEAU",
    },
    {
      "from": "ALTUS_PLATEAU",
      "metadata": {
        "requirements": {},
      },
      "to": "ALTUS_PLATEAU_LEYNDELL_OUTSKIRTS",
    },
    {
      "from": "ALTUS_PLATEAU_LEYNDELL_OUTSKIRTS",
      "metadata": {
        "description": "Leyndell turns to ash after defeating Maliketh",
        "requirements": {
          "requiredEnabledFlags": [
            "LEYNDELL_CAPITAL_ASHEN",
          ],
        },
      },
      "to": "LEYNDELL_ASHEN_CAPITAL",
    },
    {
      "from": "LEYNDELL_ASHEN_CAPITAL",
      "metadata": {
        "requirements": {
          "requiredBosses": [
            "SIR_GIDEON_OFNIR_THE_ALL_KNOWING",
          ],
        },
      },
      "to": "LEYNDELL_ASHEN_ELDEN_THRONE",
    },
    {
      "from": "LEYNDELL_ASHEN_ELDEN_THRONE",
      "metadata": {
        "requirements": {
          "requiredBosses": [
            "GODFREY_HOARAH_LOUX",
          ],
        },
      },
      "to": "LEYNDELL_FRACTURED_MARIKA",
    },
  ],
  "reasons": [],
}
`;

exports[`complex connections > Limgrave -> Somberstone Miner's Bell Bearing (5) 1`] = `
{
  "isError": false,
  "pathSteps": [
    {
      "from": "LIMGRAVE",
      "metadata": {
        "description": "Take the shortcut near Stormveil Castle",
        "requirements": {},
      },
      "to": "LIURNIA",
    },
    {
      "from": "LIURNIA",
      "metadata": {
        "description": "Access Altus Plateau via Grand Lift of Dectus",
        "requirements": {
          "requiredItems": [
            "DECTUS_MEDALLION",
          ],
        },
      },
      "to": "ALTUS_PLATEAU",
    },
    {
      "from": "ALTUS_PLATEAU",
      "metadata": {
        "requirements": {},
      },
      "to": "ALTUS_PLATEAU_LEYNDELL_OUTSKIRTS",
    },
    {
      "from": "ALTUS_PLATEAU_LEYNDELL_OUTSKIRTS",
      "metadata": {
        "description": "Enter Leyndell after defeating two Shardbearers",
        "requirements": {
          "requiredBosses": [
            "SHARDBEARERS",
            "DRACONIC_TREE_SENTINEL",
          ],
          "requiredDisabledFlags": [
            "LEYNDELL_CAPITAL_ASHEN",
          ],
        },
      },
      "to": "LEYNDELL_ROYAL_CAPITAL",
    },
    {
      "from": "LEYNDELL_ROYAL_CAPITAL",
      "metadata": {
        "requirements": {
          "requiredDisabledFlags": [
            "LEYNDELL_CAPITAL_ASHEN",
          ],
        },
      },
      "to": "ALTUS_PLATEAU_FORBIDDEN_LANDS",
    },
    {
      "from": "ALTUS_PLATEAU_FORBIDDEN_LANDS",
      "metadata": {
        "description": "Access via Grand Lift of Rold",
        "requirements": {
          "requiredItems": [
            "ROLD_MEDALLION",
          ],
        },
      },
      "to": "MOUNTAINTOPS_OF_THE_GIANTS",
    },
    {
      "from": "MOUNTAINTOPS_OF_THE_GIANTS",
      "metadata": {
        "description": "Transported to Farum Azula after burning the Erdtree",
        "requirements": {
          "requiredBosses": [
            "FIRE_GIANT",
          ],
        },
      },
      "to": "FARUM_AZULA",
    },
    {
      "from": "FARUM_AZULA",
      "metadata": {
        "requirements": {
          "requiredBosses": [
            "GODSKIN_DUO",
          ],
        },
      },
      "to": "FARUM_AZULA_POST_GODSKIN_DUO",
    },
    {
      "from": "FARUM_AZULA_POST_GODSKIN_DUO",
      "metadata": {
        "description": "Down the elevator from the 'Beside the Great Bridge' Site of Grace in the small church",
        "requirements": {},
      },
      "to": "SOMBERSTONE_MINERS_BELL_BEARING_5",
    },
  ],
  "reasons": [],
}
`;

exports[`complex connections > Limgrave -> Triple Rings of Light 1`] = `
{
  "isError": false,
  "pathSteps": [
    {
      "from": "LIMGRAVE",
      "metadata": {
        "description": "Take the shortcut near Stormveil Castle",
        "requirements": {},
      },
      "to": "LIURNIA",
    },
    {
      "from": "LIURNIA",
      "metadata": {
        "description": "Access Altus Plateau via Grand Lift of Dectus",
        "requirements": {
          "requiredItems": [
            "DECTUS_MEDALLION",
          ],
        },
      },
      "to": "ALTUS_PLATEAU",
    },
    {
      "from": "ALTUS_PLATEAU",
      "metadata": {
        "requirements": {},
      },
      "to": "ALTUS_PLATEAU_LEYNDELL_OUTSKIRTS",
    },
    {
      "from": "ALTUS_PLATEAU_LEYNDELL_OUTSKIRTS",
      "metadata": {
        "requirements": {
          "requiredGlitch": {
            "description": "The setup for this zip is a little complicated, so watching a YouTube video is advised",
            "glitch": "ZIP",
          },
        },
      },
      "to": "CONSECRATED_SNOWFIELD",
    },
    {
      "from": "CONSECRATED_SNOWFIELD",
      "metadata": {
        "requirements": {},
      },
      "to": "CONSECRATED_SNOWFIELD_ORDINA_LITURGICAL_TOWN",
    },
    {
      "from": "CONSECRATED_SNOWFIELD_ORDINA_LITURGICAL_TOWN",
      "metadata": {
        "description": "Complete the Evergaol seal puzzle to access the portal. The Sentry's Torch can be used to see hidden enemies",
        "requirements": {},
      },
      "to": "MIQUELLAS_HALIGTREE",
    },
    {
      "from": "MIQUELLAS_HALIGTREE",
      "metadata": {
        "requirements": {
          "requiredBosses": [
            "LORETTA_KNIGHT_OF_THE_HALIGTREE",
          ],
        },
      },
      "to": "ELPHAEL_BRACE_OF_THE_HALIGTREE",
    },
    {
      "from": "ELPHAEL_BRACE_OF_THE_HALIGTREE",
      "metadata": {
        "description": "In a sealed treasure room just under the Prayer Room that requires a Stonesword Key to unlock.",
        "requirements": {},
      },
      "to": "TRIPLE_RINGS_OF_LIGHT",
    },
  ],
  "reasons": [],
}
`;
